options.go

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ringpop
+
+import (
+	"io/ioutil"
+
+	"github.com/Sirupsen/logrus"
+	log "github.com/uber-common/bark"
+	"github.com/uber/ringpop-go/shared"
+)
+
+type Configuration struct {
+	// App is the name used to uniquely identify members of the same ring.
+	// Members will only talk to other members with the same app name. Note
+	// that App is taken as an argument of the Ringpop constructor and not a
+	// configuration option. This is to prevent accidental misconfiguration.
+	App string
+	// ???
+	ReplicaPoints int
+}
+
+// Options are modifiers that configure/modify a real Ringpop object
+type Option func(*Ringpop) error
+
+// applyOptions applies runtime configuration options to the specified Ringpop
+// instance.
+func applyOptions(r *Ringpop, opts []Option) error {
+	for _, option := range opts {
+		err := option(r)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Runtime options
+
+func Channel(ch shared.TChannel) Option {
+	return func(r *Ringpop) error {
+		r.channel = ch
+		return nil
+	}
+}
+
+func Logger(l log.Logger) Option {
+	return func(r *Ringpop) error {
+		r.logger = l
+		r.log = l
+		return nil
+	}
+}
+
+func ReplicaPoints(n int) Option {
+	return func(r *Ringpop) error {
+		r.config.ReplicaPoints = n
+		return nil
+	}
+}
+
+func Statter(s log.StatsReporter) Option {
+	return func(r *Ringpop) error {
+		r.statter = s
+		return nil
+	}
+}
+
+// IdentityResolver is a function that returns the listen interface/port
+// that Ringpop should identify as.
+type IdentityResolver func() (string, error)
+
+func IdentityResolverFunc(resolver IdentityResolver) Option {
+	return func(r *Ringpop) error {
+		r.identityResolver = resolver
```

Comment:
We should be careful about nils. In theory, a user can pass in a nil `resolver`, right?